snapcraft/internal/pluginhandler/__init__.py

```diff
@@ -1059,38 +1131,157 @@ def _file_collides(file_this, file_other):
+                new_part_files.add(fileset[part_file])
+            else:
+                new_part_files.add(part_file)
 
-            if conflict_files:
-                raise SnapcraftPartConflictError(
-                    other_part_name=other_part_name,
-                    part_name=part.name,
-                    conflict_files=conflict_files)
+        # map new names to old names
+        filemap = {}
+        for part_file in new_part_files:
+            filemap[part_file] = rev_fileset.get(part_file, part_file)
+
+        _check_previous_parts_for_collisions(
+            parts_files, new_part_files, part, rev_fileset)
 
         # And add our files to the list
-        parts_files[part.name] = {'files': part_files,
+        parts_files[part.name] = {'files': new_part_files,
+                                  'filemap': filemap,
                                   'installdir': part.installdir}
 
 
+def _get_path_prefixes(filepath):
+    dirname = os.path.dirname(filepath)
+    result = []
+
+    while dirname != '/' and dirname != '':
+        result.append(dirname)
+        dirname = os.path.dirname(dirname)
+
+    return result
+
+
+def _add_filepath_to_fileset(filepath, srcdir, fileset, dirset):
+    if type(filepath) == tuple:
+        path = filepath[0]
+    else:
+        path = filepath
+
+    if os.path.isdir(os.path.join(srcdir, path)):
+        dirset.add(filepath)
+    else:
+        fileset.add(filepath)
+
+    return fileset, dirset
+
+
+def _organize_filepath_prefixes(filepath, rev_organize_fileset,
+                                organize_fileset, srcdir, fileset, dirs):
+
+    path_prefixes = _get_path_prefixes(filepath)
+    matched = False
+    for prefix in path_prefixes:
+        if prefix in rev_organize_fileset and filepath.startswith(prefix):
+            for rev in rev_organize_fileset[prefix]:
+                new_filepath = ("{}{}".format(rev,
+                                filepath[len(prefix):]),
+                                filepath)
+                fileset, dirs = _add_filepath_to_fileset(
+                    new_filepath, srcdir, fileset, dirs)
+            matched = True
+            # XXX: This seems weird it's the last prefix path that matched.
+            filepath = new_filepath
+            break
+        elif prefix in organize_fileset and filepath.startswith(prefix):
+            new_filepath = (
+                filepath,
+                "{}{}".format(organize_fileset[prefix],
+                              filepath[len(prefix):]))
+            fileset, dirs = _add_filepath_to_fileset(
+                new_filepath, srcdir, fileset, dirs)
+            matched = True
+            filepath = new_filepath
+            break
+
+    if matched:
+        fileset, dirs = _add_filepath_to_fileset(
+            filepath, srcdir, fileset, dirs)
+    else:
+        fileset.add(filepath)
+
+    return fileset, dirs
+
+
+def _organize_fileset(fileset_orig, organize_fileset, srcdir):
```

Comment:
I'm having some serious problems understanding what's happening through here. Combined with the fact that the pluginhandler is already quite bloated, I'm starting to think this behavior would be better suited to being extracted into another module, perhaps a class. And documented very well.